### DISCUSSION TITLE
Need to prevent convert from "upgrading" bitmap from 1-bit to 8-bit when using the -resize option

### DISCUSSION THREAD

**Discussion by smt62** [ImageMagick/ImageMagick#6902] [Help]
### ImageMagick version

6.9.12-64 Q16 x86_64 17467

### Operating system, version and so on

Oracle Linux Server release 8.6

### Description

When I use the following command: _convert MyBmp.bmp -resize 450x150 MyNewBmp.bmp_ the resulting bitmap file is "upgraded" from 1-bit to 8-bit.  Need help with two things.  First, how can I prevent convert from "upgrading" the images from 1 to 8-bit?  Second, is there a way to suppress adherence to the aspect ratio (and generate an image with the geometry specified) of the original image when using the -resize option with convert?

Thank you.

**Comment by fmw42** [comment]
-resize defaults to interpolation when resizing and that introduce new pixel values such that there are more than 1-bit values.

Use 

`-filter point -resize 450x150`

**Reply by smt62** [reply]
Hi Fred,

Thank you for your quick reply!  The -filter point parm retains the 1-bit.  the "450x150!" option ignores aspect ration.  So, getting closer.

Now, the only remaining issue is the quality of the resulting image.

![sidebyside](https://github.com/ImageMagick/ImageMagick/assets/50673356/5e020535-c203-4f30-87f0-c58b2dcde05a)

The image on the left is the original before conversion.  The one on the right is post conversion.  The quality is not acceptable for our purposes.  Barely appears when printed.

I would have expected quality degradation if I was enlarging the image because the software has to make up/interpolate data to fill the extra space/size of the image.

I did not expect degradation when shrinking a image.

Are there other parms/options that will do this shrink without degrading image quality?  Other utilities?  Other image formats that are "lossless" during conversion?

I'm not an image guy.  Just a lowly application programmer.

Thoughts?

Scott

**Reply by fmw42** [reply]
Please post the original image, not a screen snap.

**Reply by smt62** [reply]
Hi Fred,

Github won't allow me to attach BMPs.

I could convert it to another format but that would defeat the purpose wouldn't it?

Scott

**Reply by fmw42** [reply]
Just zip the BMP and it will attach

**Reply by smt62** [reply]
[SignatureBmps.zip](https://github.com/ImageMagick/ImageMagick/files/13531729/SignatureBmps.zip)

Images attached in zip.  800x300 was starting point

**Comment by snibgo** [comment]
To resize to exact dimensions, use `!`:
```
-resize "450x150!"
```

**Comment by snibgo** [comment]
> I did not expect degradation when shrinking a image.

Your expectation is doomed to fail. Your black line is only one pixel wide. If you reduce it by 50%, it will be 0.5 pixels wide, which would make the pixels 50% gray. But you also specify that you want only black or white only, so we would expect that 50% of the pixels on the line will become white, and 50% will be black.

We can fix this by doing an ordinary resize, followed by a threshold.
```
magick 800x300.bmp -resize "450x150!" -threshold 75% x.png
```

![x](https://github.com/ImageMagick/ImageMagick/assets/33812368/b46fe5e6-ed97-481c-b511-ad78dcd5abe7)

An alternative is to dilate the black line before `-filter point -resize 450x150!`.

**Reply by smt62** [reply]
So, this worked perfectly to generate a .png.  Only 725 bytes!  Why are .png's so small?  Was amazed.  When I use same command to generate a .bmp, two things happen.  First, size balloons from ~30K to ~203K.  Second is that once  again, convert "upgrades" the image to an 8-bit image which I do not want. 

Oddly enough, the following two commands combined yield what I think (can't test on printer until tomorrow AM) I want.

convert MySIGFromReprint.bmp -resize "450x150!" -threshold 75% x.png
Running identify against x.png yields:  x.png PNG 450x150 450x150+0+0 8-bit Gray 2c 725B 0.000u 0:00.000

convert x.png  xx.bmp
Running identify against xx.bmp yields: xx.bmp BMP 450x150 450x150+0+0 1-bit sRGB 2c 9130B 0.000u 0:00.000
 
Is there a command/utility that would take me directly from a vector graphics (.svg/.vg) file to xx.bmp (1-bit, relatively small size)?

**Comment by snibgo** [comment]
You can get directly to a 1-bit BMP like this:
```
magick 800x300.bmp -resize "450x150!" -threshold 75% -type palette x.bmp
```
This creates a colormap (palette) with 2 entries, so it encodes as 1 bit per pixel.

> Is there a command/utility that would take me directly from a vector graphics (.svg/.vg) file to xx.bmp (1-bit, relatively small size)?

If your SVG raterizes to more than two colours, you will need to reduce them.

**Reply by smt62** [reply]
First of all, thank you for your time.

This command: _magick 800x300.bmp -resize "450x150!" -threshold 75% -type palette x.bmp_ generates the image size/quality I want. But, when I send it to the Epson receipt printer, instead of coming out as black text on white paper, it is coming out as a black background box with white text in the box.  Totally legible.  Quality's good but it looks like a photo negative.  Thoughts?

**Reply by smt62** [reply]
![Sig](https://github.com/ImageMagick/ImageMagick/assets/50673356/d88b2823-320a-43ee-add5-8eb5e0e3e36a)

This is what I mean.  Ignore that the pic is grainy.  Quality is fine for my purposes.  Just don't want it looking like a photo negative.  Should be black text on a white (or clear/no) background

**Comment by snibgo** [comment]
Perhaps the printer is confusing RGB with CMYK. Printer driver software can normally accept RGB data, and convert that to CMYK as required.

BMP files always contain RGB channels, where all-zeros represent black, and maximum values represent white. CMYK is the opposite. I suppose you could insert `-negate`:
```
magick 800x300.bmp -resize "450x150!" -threshold 75% -negate -type palette x.bmp 
```

**Reply by smt62** [reply]
This worked.  Thank you! Looking very good.  Now I'm getting greedy.  Is there any way to thicken/darken the text of the signature?  If not, no worries.

**Comment by snibgo** [comment]
I mentioned above that "An alternative is to dilate the black line". We do this by eroding the white:
```
magick 800x300.bmp -morphology erode disk -resize "450x150!" x.png
```

![x](https://github.com/ImageMagick/ImageMagick/assets/33812368/2cb7ddec-7013-455b-acbf-951f471fd630)

We can use `disk:1.0` for a little thickening, or `disk:5.3` for heavy thickening, etc. See https://imagemagick.org/Usage/morphology/#disk

**Reply by smt62** [reply]
The  ...erode disk ... command generates nice output but the epson receipt printer does not like it.  So, I'll stick with the "unmorphed" version.

Is there a single magick command that will go directly from a vector graphics file to a .png or .bmp? Or does it require a series of commands?

**Comment by snibgo** [comment]
If the vector graphics file is SVG or AI or PDF or some other:
```
magick in.svg out.bmp
```
IM delegates AI and PDF inputs to Ghostscript. IM may delegate SVG inputs to Inkscape.
